meshroom/ui/components/thumbnail.py

```diff
@@ -0,0 +1,210 @@
+from meshroom.common import Signal
+
+from PySide2.QtCore import QObject, Slot, QSize, QUrl, Qt
+from PySide2.QtGui import QImageReader, QImageWriter
+
+import os
+from pathlib import Path
+import hashlib
+import time
+import logging
+from multiprocessing.pool import ThreadPool
+
+
+class ThumbnailCache(QObject):
+    """ThumbnailCache provides an abstraction for the thumbnail cache on disk, available in QML.
+
+    For a given image file, it ensures the corresponding thumbnail exists (by creating it if necessary)
+    and gives access to it.
+    Since creating thumbnails can be long (as it requires to read the full image from disk)
+    it is performed asynchronously to avoid blocking the main thread.
+
+    The default cache location is a subdirectory of the user's home directory:
+    ~/Meshroom/thumbnails.
+    This location can be overriden with the MESHROOM_THUMBNAIL_DIR environment variable.
+
+    This class also takes care of cleaning the thumbnail directory,
+    i.e. scanning this directory and removing thumbnails that have not been used for too long.
+    This operation also ensures that the number of thumbnails on disk does not exceed a certain limit, 
+    by removing thumbnails if necessary (from least recently used to most recently used).
+
+    The default time limit is 90 days, 
+    and can be overriden with the MESHROOM_THUMBNAIL_TIME_LIMIT environment variable.
+
+    The default maximum number of thumbnails on disk is 100000, 
+    and can be overriden with the MESHROOM_MAX_THUMBNAILS_ON_DISK.
+
+    The main use case for thumbnails in Meshroom is in the ImageGallery.
+    """
+
+    # Thumbnail cache directory
+    thumbnailDir = os.path.join(Path.home(), 'Meshroom', 'thumbnails')
+
+    # Thumbnail dimensions limit (the actual dimensions of a thumbnail will depend on the aspect ratio)
+    thumbnailSize = QSize(100, 100)
+
+    # Time limit for thumbnail storage on disk, expressed in days
+    storageTimeLimit = 90
+
+    # Maximum number of thumbnails in the cache directory
+    maxThumbnailsOnDisk = 100000
+
+    # Signal to notify listeners that a thumbnail was created and written on disk
+    # This signal has one argument: the url of the image that the thumbnail is associated to
+    thumbnailCreated = Signal(QUrl)
+
+    # Thread pool for running createThumbnail asynchronously on a fixed number of worker threads
+    pool = ThreadPool()
```

Comment:
We need to limit the number of processes (by default will use the number of cpus on the workstation) as loading too many images at the same time will not improve the overall time and will slow down the load of the first image.
```suggestion
    pool = ThreadPool(processes=3)
```